### PR TITLE
Components: Make the Popover.Slot optional

### PR DESCRIPTION
**Pull request by youknowriad**
Related #53874

## What and why?

Gutenberg can be used as a platform/framework to build block editors. Mainly thanks to the @wordpress/block-editor package. That said, the experience today is not as straightforward as it can be. There can be a lot of small gotchas and hacks you need to do in order to achieve the desired result. One of these small things is the need to manually render `Popover.Slot` somewhere in the React tree.

I think ideally, Popover components should work properly out of the box without any provider or slot, and only if we really want to control their position in the DOM, we should use `Popover.Slot`.

This PR does that by using an "automatic container" for all popovers that is appended to the end of the "body" of the document if there's no slot that is defined manually using `Popover.Slot`. I think that's a better default than inlining popovers which breaks all the styles and more.

**Todo:**

 - [x] I've noticed that when the anchor is within an iframe, the popover is not offset properly to the position of the iframe (you can test for instance the appender of the navigation block). I'm not entirely sure how to fix that because I thought we had that offsetting in place already. cc @talldan @ciampo 

## Testing Instructions

1- Check that the stories updates (removal of popover and slot provider) still work properly.
2- Navigate the editor and ensure popovers, tooltips, drodowns are shown properly.

### CODE REVIEW

File: packages/block-editor/src/components/url-popover/test/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/url-popover/test/index.js:1]: I'm removing this test because IMO, this test is just useless. It tests just the snapshot of the elements which IMO is just internal implementation details that will change every time we change how the component renders.

File: packages/block-editor/src/components/block-popover/index.js
Comment (jsnajdr) [inline packages/block-editor/src/components/block-popover/index.js]: What went wrong with this particual popover when it wasn't rendered inline? I'm wondering if the "in-between" inserters shouldn't also remain inline.

If the inline rendering continues to be a valid and frequent use case, it would deserve being a prop: `<Popover inline={ true } />`

Comment (youknowriad) [reply packages/block-editor/src/components/block-popover/index.js]: yeah, we have some tests about "tab order" and things like that. Basically this one needs to be rendered before the block list.

To be honest, it seem that there's so much history here, with all the unstable props and popover slots. I think this one probably deserves its own dedicated PR to clear things out. @ellatrix would know more about this one.

---

I didn't check the in between inserters yet.

Comment (youknowriad) [reply packages/block-editor/src/components/block-popover/index.js]: I hesitated about the "inline" prop but I added it, it seems a bit simpler than having to add slots everywhere.

**Comment by youknowriad** [comment]
Ok, so I've just realized that trunk also has the "navigation" dropdown offset, so it's not related to iframing or the current PR. Meaning this one should be ready to go. :)

File: packages/block-editor/src/components/block-popover/index.js
Comment (jsnajdr) [reply packages/block-editor/src/components/block-popover/index.js]: There is some ambiguity when there is a `Popover.__unstableSlotNameProvider` together with `inline` prop. Example:
```js
function MyPopover( { __unstablePopoverSlot } ) {
  return <Popover
    __unstableSlotName={ __unstablePopoverSlot }
    inline={ ! __unstablePopoverSlot }
  />;
}

function App() {
  return (
    <Popover.__unstableSlotNameProvider value="block-tools-after">
      <MyPopover />
    </Popover.__unstableSlotNameProvider>
  );
}
```
Here the `App`'s intent was to render the `MyPopover` in the `block-tools-after` slot, but something else will happen: it will be rendered inline, because the `inline` prop is `true`.

It comes down to whether `inline={ true }` means
- prefer inline: if there is no slot either passed as a prop or in context, render the popover inline rather than in the fallback portal.
- force inline: render the popover always inline, even though a slot is specified. Ignore the slot.

~The current implementation is closer to "prefer inline", but not exactly: it doesn't take into account the slot in the context.~

The context provider is used by the `RichText` component to render `FormatEdit` popovers.

Comment (youknowriad) [reply packages/block-editor/src/components/block-popover/index.js]: I'd go with "force inline" personally to avoid ambiguity.

Comment (youknowriad) [reply packages/block-editor/src/components/block-popover/index.js]: For me the code is forcing inline regardless of the context or the slot name prop and this is what I wanted to implement. So I'm not sure I understand.

Comment (jsnajdr) [reply packages/block-editor/src/components/block-popover/index.js]: I'm sorry, I misread the `shouldRenderWithinSlot = slot.ref && ! inline` condition.

When forcing inline, however, the `inline={ ! __unstablePopoverSlot }` prop in `BlockPopover` and `BlockPopoverInbetween` is wrong, as it's forcing inline when a slot is provided by context.

In practice this probably never happens, if the `Popover.__unstableSlotNameProvider` is really used only by `RichText`. `RichText` attaches its popovers to the edited content, it doesn't know what a block is. So it will never be used together with `BlockPopover`.

**Review by jsnajdr — APPROVED** [review]
Everything seems OK when I'm testing this change.

There is incompatibility between `BlockPopover` and `SlotNameProvider` described in https://github.com/WordPress/gutenberg/pull/53889#discussion_r1304068708, but in practice these two pieces of code never actually meet each other.

File: packages/block-editor/src/components/block-popover/index.js
Comment (youknowriad) [reply packages/block-editor/src/components/block-popover/index.js]: To be honest, I missed when `__unstableSlotNameProvider` was introduced and it feels a bit weird to me, I'll have to dive there to see whether this is still something we want.

File: packages/components/CHANGELOG.md
Comment (tyxla) [inline packages/components/CHANGELOG.md:5]: Isn't this a breaking change?

File: packages/components/src/popover/index.tsx
Comment (tyxla) [inline packages/components/src/popover/index.tsx:583]: Nit: could be a single condition:

```suggestion
	if ( shouldRenderWithinSlot || inline ) {
		return content;
	}
```

File: packages/components/src/popover/README.md
Comment (tyxla) [inline packages/components/src/popover/README.md:9]:
```suggestion
Render a Popover adjacent to its container.
```

**Comment by jsnajdr** [comment]
> I've noticed that when the anchor is within an iframe, the popover is not offset properly to the position of the iframe (you can test for instance the appender of the navigation block).

Could you share a screenshot of what's happening there? The iframe positioning is undergoing major changes in #46845, so the Navigation case would be worth testing with that PR applied.

File: packages/components/src/popover/index.tsx
Comment (tyxla) [inline packages/components/src/popover/index.tsx:145]: Nit of the nits: the classname could be declared in a constant so we don't have to repeat it multiple times.

File: packages/block-editor/src/components/url-popover/test/index.js
Comment (tyxla) [reply packages/block-editor/src/components/url-popover/test/index.js:1]: Agreed!

**Comment by youknowriad** [comment]
Screenshot 2023-08-24 at 11 55 30 AM

As you can see on the screenshot above, the popover is a bit far from the "plus" button that I clicked. I initially thought it was because of the iframe offset (surprisingly the offset seen on the screenshot is almost equal to the offset of the iframe) but I ultimately figured that the iframe offset is working properly but that for some reason, this particular popover has an extra offset applied (not sure where) because in trunk it's the same.

Then after testing on the site editor with the navigation menu on the top, I realized that the block toolbar appears below the block when the navigation is at the top of the template so I'm guessing that this extra space/offset is intentional to prevent the toolbar from hiding the dropdown. Probably not the best solution, but at least it's not impacted by this PR.

File: packages/components/src/popover/index.tsx
Comment (tyxla) [inline packages/components/src/popover/index.tsx:184]: Would be nice to add some tests for this one.

**Comment by tyxla** [comment]
Would have been great to get a review from the components folks too IMHO.

File: packages/components/CHANGELOG.md
Comment (youknowriad) [reply packages/components/CHANGELOG.md:5]: If you were using `Popover.Slot` before, it would continue to work.

Some popovers might not render "inline" while they previously did, which could have subtle impacts but I'm not sure it's big enough to consider this a breaking change.

File: packages/components/src/popover/index.tsx
Comment (youknowriad) [reply packages/components/src/popover/index.tsx:583]: Right. Sorry the PR was auto-merged. I'll open a follow up with some of these changes.

**Comment by jsnajdr** [comment]
> As you can see on the screenshot above, the popover is a bit far from the "plus" button that I clicked.

I think this is a bug, and the Floating UI update in #46845 fixes it. The popover should be aligned right under the "Add link" text. The problem is that we inadverently already upgraded the `@floating-ui/dom` library to version 1.1+, a version that contains the native iframe positioning. This new feature was released as a minor version bump. We got the new version probably as a side-effect of adding Ariakit, or something similar. And now the iframe offset is added twice, both natively and in our custom middleware.

**Comment by brookewp** [comment]
I believe this is related to the positioning issue @youknowriad mentioned above, but the problem is obvious in the '[Toggle Popover](https://wordpress.github.io/gutenberg/?path=/story/components-popover--default)' examples in Storybook. The popover is rendering but is positioned out of view:

Screen Capture on 2023-08-25 at 10-21-42

I tested this against the PR @jsnajdr mentioned above but am encountering the same result, where the examples in Storybook appear broken after this PR merge.

**Comment by youknowriad** [comment]
Thanks for the heads up, I will take a look

**Comment by jsnajdr** [comment]
> The popover is rendering but is positioned out of view:

I don't have a solution yet, just an observation: before this PR, the Storybook popover was rendered inline inside the "Toggle Popover" button. Now it's rendered inside a portal element that's a direct child of the document's body.

Seems to me that the Floating UI positioning algorithm runs into troubles in a situation where the reference (anchor) element is in an element that is scrollable (`overflow: auto`) and is actually scrolled. Then the positioning doesn't account for the scroll position.

The Storybook page has a large (300vw x 300vh) scrollable "canvas" element where the "Toggle Popover" button is rendered in its center, and there is an `useEffect( () => { buttonEl.scrollIntoView() } )` call that scrolls the canvas to make it visible.

In the PR (#46845) that upgrades Floating UI to latest version and removes all custom iframe positioning, I see a slightly broken behavior in the Site Editor "zoom out mode". And there is also a scrolling element there!

**Comment by youknowriad** [comment]
The problem was that the "anchor" was not rendered at the expected position. It should be fixed by #53982

File: packages/components/CHANGELOG.md
Comment (danieliser) [reply packages/components/CHANGELOG.md:5]: Definitely was, previously you could define styles taking advantage of the inline setup, but now CSS selectors need to be rewritten.

**Comment by georgestephanis** [comment]
👋🏼 Just reporting in that I bisected an issue we were having down to f7741a9a9703dc4e1f3365e277da070d1a6dfd2b in this PR --

Before the pr -- up to version 16.5, we were doing some pop-up calendars with a Team51 `simple-events` plugin that looked like this:

Screenshot 2023-09-27 at 3 39 30 PM

but after, a lot of the styling that would ordinarily propagate through the Emotion js/css library is missing, leading them to look like this:

Screenshot 2023-09-27 at 3 41 58 PM

When I check the source, when it's behaving, the emotion style attributes seem to line up correctly like this:

Screenshot 2023-09-26 at 3 47 14 PM
Screenshot 2023-09-26 at 3 46 45 PM

But after this goes in, they don't match properly, so the styles don't seem to apply as they aught:

Screenshot 2023-09-27 at 2 29 44 PM
Screenshot 2023-09-27 at 2 29 24 PM

I'm a bit out of my league here, but any guidance on if it's an upstream issue, or if we're implementing something incorrectly?  I'm happy to add anyone to our plugin repo for direct code access if it can help illuminate the issue.

**Comment by ciampo** [comment]
Hey @georgestephanis , you should be able to get back to the same `Popover` behaviour as before, as @youknowriad also explained in https://github.com/WordPress/gutenberg/pull/53982#issuecomment-1720834686:

> Just noting that regardless of what you're trying to do, it's possible to have the exact behavior as before:
> 
> * If you want to render the popover inline you can just pass the "inline" prop.
> * If you want to render the popover in a specific position you can use the Popover.Slot component.

Hope this helps!

**Comment by georgestephanis** [comment]
Thanks for the pointer, @ciampo!  I'm making some progress with it -- but the code we've got is using `Dropdown`, rather than `Popover` directly -- so if I add in `popoverProps={ { inline: 'inline' } }` to the `Dropdown` I get something ... close ... ish.  The datepicker table styles seem to propagate as they should, but it seems there's some other interfering styles as well --

Screenshot 2023-09-27 at 7 44 19 PM

I assume it's likely because of where in the hierarchy it is rendering inline, rather than where it used to, higher up in an autogenerated popover slot.

Screenshot 2023-09-27 at 7 46 31 PM

Is there an example somewhere that I can work off of to recreate the higher slot based rendering to match the prior behavior?

**Comment by tyxla** [comment]
Haven't double-checked, but maybe we just missed applying the `inline` prop to the `Popover` instance in the `Dropdown` component?

**Comment by youknowriad** [comment]
>  so if I add in popoverProps={ { inline: 'inline' } } to the Dropdown I get something ... close ... ish. The datepicker table styles seem to propagate as they should, but it seems there's some other interfering styles as well --

I think if you apply `popoverProps={ { inline: true } }` to the Dropdown, we're supposed to get the exact same DOM position as before, so the styles should be the same. Maybe there's something else at play here.

File: packages/components/src/popover/index.tsx
Comment (mcsf) [inline packages/components/src/popover/index.tsx:144]: @youknowriad: At the risk of asking a stupid question, does this reliance on a global document carry any consequences for usages of Popover within/across the iframe boundary of block editors?

Comment (youknowriad) [reply packages/components/src/popover/index.tsx:144]: no, I don't think there's impact on the iframe slots here. This one is the fallback if there's no Slot defined while if we had to render something within the iframe previously we were forced to have a Slot.

Comment (ciampo) [reply packages/components/src/popover/index.tsx:144]: Maybe to be extra thorough, we could use the anchor or the floating's `ownerDocument`?

**Pull request merged**
